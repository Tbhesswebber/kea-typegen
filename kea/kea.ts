// This will be integrated into Kea soon... It's just faster to iterate here.

import { kea as regularKea, useValues as useRegularValues } from 'kea'

type BlankLogic = {
    actions?: any
    actionCreators?: any
    actionTypes?: any
    reducers?: any
    selectors?: any
    listeners?: any
    values?: any
    __selectorTypeHelp?: any
}

type ReducerActions<LogicType extends BlankLogic, ReducerType> = {
    [K in keyof LogicType['actions']]?: (
        state: any,
        payload: ReturnType<LogicType['actions'][K]>['payload'],
    ) => ReducerType
}
type ReducerDefinitions<LogicType extends BlankLogic> = {
    [K in keyof LogicType['reducers']]?:
        | [ReturnType<LogicType['reducers'][K]>, ReducerActions<LogicType, ReturnType<LogicType['reducers'][K]>>]
        | ReducerActions<LogicType, ReturnType<LogicType['reducers'][K]>>
}
type Selector = (state?: any, props?: any) => any

type SelectorDefinition<Selectors, S extends Selector, SelectorFunction extends any> = [
    (
        s: Selectors,
    ) =>
        | []
        | [Selector]
        | [Selector, Selector]
        | [Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector],
    SelectorFunction,
]

type SelectorDefinitions<LogicType extends BlankLogic> = {
    [K in keyof LogicType['selectors']]?: SelectorDefinition<
        LogicType['selectors'],
        LogicType['selectors'][K],
        LogicType['__selectorTypeHelp'][K]
    >
}

type ListenerDefinitions<LogicType extends BlankLogic> = {
    [K in keyof LogicType['actions']]?:
        | ((
              payload: ReturnType<LogicType['actions'][K]>['payload'],
              breakpoint: (() => void) | ((ms: number) => Promise<void>),
              action: ReturnType<LogicType['actions'][K]>,
              previousState: any,
          ) => void | Promise<void>)
        | (() => void | Promise<void>)
}
type LogicInput<LogicType> = {
    path?: any
    constants?: any
    actions?: (logic: LogicType) => any | any
    reducers?: ReducerDefinitions<LogicType> | ((logic: LogicType) => ReducerDefinitions<LogicType>)
    events?: (logic: LogicType) => any | any
    selectors?: SelectorDefinitions<LogicType> | ((logic: LogicType) => SelectorDefinitions<LogicType>)
    listeners?: ListenerDefinitions<LogicType> | ((logic: LogicType) => ListenerDefinitions<LogicType>)
    loaders?: any
}

export function kea<LogicType>(input: LogicInput<LogicType>): LogicType {
    return regularKea(input)
}
export function useValues<L extends BlankLogic>(logic: L): L['values'] {
    return useRegularValues(logic)
}
