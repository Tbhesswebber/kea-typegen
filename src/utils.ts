import * as ts from 'typescript'
import { visitProgram } from './visit/visit'
import { parsedLogicToTypeString } from './print/print'

export function logicSourceToLogicType(logicSource: string) {
    const program = programFromSource(logicSource)
    const [parsedLogic] = visitProgram(program)
    return parsedLogicToTypeString(parsedLogic)
}

export function sourceToSourceFile(sourceCode: string, filename: string = 'logic.ts') {
    return ts.createSourceFile(filename, sourceCode, ts.ScriptTarget.ES5, true)
}

export function programFromSource(sourceCode: string) {
    const options = {}
    const compilerHost = ts.createCompilerHost(options)
    compilerHost.getSourceFile = (fileName) => (fileName === 'logic.ts' ? sourceToSourceFile(sourceCode) : undefined)
    return ts.createProgram(['logic.ts'], options, compilerHost)
}

export function isKeaCall(node: ts.Node, checker: ts.TypeChecker) {
    if (!ts.isIdentifier(node)) {
        return false
    }

    if (!ts.isCallExpression(node.parent)) {
        return false
    }

    if (!node.parent || !ts.isCallExpression(node.parent)) {
        return false
    }

    const symbol = checker.getSymbolAtLocation(node)
    if (!symbol || symbol.getName() !== 'kea') {
        return false
    }

    const input = node.parent.arguments[0]

    if (!ts.isObjectLiteralExpression(input)) {
        return false
    }

    return true
}

export function getTypeNodeForDefaultValue(defaultValue: ts.Node, checker: ts.TypeChecker): ts.TypeNode {
    let typeNode
    if (defaultValue) {
        if (ts.isAsExpression(defaultValue)) {
            typeNode = defaultValue.type
            if (ts.isParenthesizedTypeNode(typeNode)) {
                typeNode = typeNode.type
            }
        } else if (
            defaultValue?.kind === ts.SyntaxKind.TrueKeyword ||
            defaultValue?.kind === ts.SyntaxKind.FalseKeyword
        ) {
            typeNode = ts.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword)
        } else if (ts.isStringLiteralLike(defaultValue)) {
            typeNode = ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword)
        } else if (ts.isNumericLiteral(defaultValue)) {
            typeNode = ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)
        } else {
            typeNode = checker.typeToTypeNode(checker.getTypeAtLocation(defaultValue))
        }
    } else {
        typeNode = ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
    }
    return typeNode
}

export function getParameterDeclaration(param: ts.ParameterDeclaration) {
    return ts.createParameter(
        undefined,
        undefined,
        undefined,
        ts.createIdentifier(param.name.getText()),
        param.initializer || param.questionToken ? ts.createToken(ts.SyntaxKind.QuestionToken) : undefined,
        param.type || ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
        undefined,
    )
}
