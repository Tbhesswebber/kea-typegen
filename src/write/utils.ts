import { ParsedLogic } from '../types'
import { visit, types } from 'recast'

export const t = types.namedTypes
export const b = types.builders

export function isKeaCall(path: any): boolean {
    const stmt = path.node
    return (
        t.Identifier.check(stmt.callee) &&
        stmt.callee.name === 'kea' &&
        stmt.arguments[0] &&
        (t.ObjectExpression.check(stmt.arguments[0]) || t.ArrayExpression.check(stmt.arguments[0])) &&
        path.parentPath &&
        t.VariableDeclarator.check(path.parentPath.value) &&
        t.Identifier.check(path.parentPath.value.id)
    )
}

export function visitAllKeaCalls(
    ast: any,
    parsedLogics: ParsedLogic[],
    filename: string,
    callback: (args: { parsedLogic: ParsedLogic; logicName: string; path: any }) => any,
): void {
    visit(ast, {
        visitCallExpression: function (path) {
            if (!isKeaCall(path)) {
                this.traverse(path)
            }
            const logicName = path.parentPath.value.id.name
            if (!logicName) {
                console.warn(
                    `[KEA-TYPEGEN] Can not add path to logic in "${filename}:${path.node.loc.start}" because it's not stored as a variable.`,
                )
                return
            }
            const parsedLogic = parsedLogics.find((l) => l.logicName === logicName)
            if (!parsedLogic) {
                console.error(
                    `[KEA-TYPEGEN] While trying to add a path, could not find logicName "${logicName}" in the list of logicNames (${Object.keys(
                        parsedLogics.map((l) => l.logicName),
                    ).join(', ')}) in the file: ${filename}`,
                )
                return
            }
            callback.bind(this)({ logicName, parsedLogic, path })
            return false
        },
    })
}
