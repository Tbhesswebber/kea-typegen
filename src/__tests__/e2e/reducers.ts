import { logicSourceToLogicType } from '../../utils'

test('reducers - with a function', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: () => ({
                updateName: (name: string) => ({ name }),
                updateOtherName: (otherName: string) => ({ otherName }),
            }),
            reducers: () => ({
                name: [
                    'my name',
                    {
                        updateName: (_, { name }) => name,
                        [actions.updateOtherName]: (state, payload) => payload.name,
                    },
                ],
                otherNameNoDefault: {
                    updateName: (_, { name }) => name,
                },
                yetAnotherNameWithNullDefault: [
                    null as (string | null),
                    {
                        updateName: (_, { name }) => name,
                    },
                ],
            })
        })
    `
    expect(logicSourceToLogicType(logicSource)).toEqual(
        `
export interface logicType {
    actionCreators: {
        updateName: (name: string) => ({
            type: "update name (logic)";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: "update other name (logic)";
            payload: { otherName: string; };
        });
    };
    actions: {
        updateName: (name: string) => ({
            type: "update name (logic)";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: "update other name (logic)";
            payload: { otherName: string; };
        });
    };
    reducer: (state: any, action: () => any, fullState: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    reducers: {
        name: (state: string, action: any, fullState: any) => string;
        otherNameNoDefault: (state: any, action: any, fullState: any) => any;
        yetAnotherNameWithNullDefault: (state: string | null, action: any, fullState: any) => string | null;
    };
    selector: (state: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    selectors: {
        name: (state: any, props: any) => string;
        otherNameNoDefault: (state: any, props: any) => any;
        yetAnotherNameWithNullDefault: (state: any, props: any) => string | null;
    };
    values: {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
}`.trim(),
    )
})

test('reducers - as an object', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: () => ({
                updateName: (name: string) => ({ name }),
                updateOtherName: (otherName: string) => ({ otherName }),
            }),
            reducers: {
                name: [
                    'my name',
                    {
                        updateName: (_, { name }) => name,
                    },
                ],
                otherNameNoDefault: {
                    updateName: (_, { name }) => name,
                },
                yetAnotherNameWithNullDefault: [
                    null as (string | null),
                    {
                        updateName: (_, { name }) => name,
                    },
                ],
            }
        })
    `
    expect(logicSourceToLogicType(logicSource)).toEqual(
        `
export interface logicType {
    actionCreators: {
        updateName: (name: string) => ({
            type: "update name (logic)";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: "update other name (logic)";
            payload: { otherName: string; };
        });
    };
    actions: {
        updateName: (name: string) => ({
            type: "update name (logic)";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: "update other name (logic)";
            payload: { otherName: string; };
        });
    };
    reducer: (state: any, action: () => any, fullState: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    reducers: {
        name: (state: string, action: any, fullState: any) => string;
        otherNameNoDefault: (state: any, action: any, fullState: any) => any;
        yetAnotherNameWithNullDefault: (state: string | null, action: any, fullState: any) => string | null;
    };
    selector: (state: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    selectors: {
        name: (state: any, props: any) => string;
        otherNameNoDefault: (state: any, props: any) => any;
        yetAnotherNameWithNullDefault: (state: any, props: any) => string | null;
    };
    values: {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
}`.trim(),
    )
})

test('reducers - as a function returning a object', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: () => ({
                updateName: (name: string) => ({ name }),
                updateOtherName: (otherName: string) => ({ otherName }),
            }),
            reducers: () => { 
                return {
                    name: [
                        'my name',
                        {
                            updateName: (_, { name }) => name,
                            [actions.updateOtherName]: (state, payload) => payload.name,
                        },
                    ],
                    otherNameNoDefault: {
                        updateName: (_, { name }) => name,
                    },
                    yetAnotherNameWithNullDefault: [
                        null as (string | null),
                        {
                            updateName: (_, { name }) => name,
                        },
                    ],
                }
            }
        })
    `
    expect(logicSourceToLogicType(logicSource)).toEqual(
        `
export interface logicType {
    actionCreators: {
        updateName: (name: string) => ({
            type: "update name (logic)";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: "update other name (logic)";
            payload: { otherName: string; };
        });
    };
    actions: {
        updateName: (name: string) => ({
            type: "update name (logic)";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: "update other name (logic)";
            payload: { otherName: string; };
        });
    };
    reducer: (state: any, action: () => any, fullState: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    reducers: {
        name: (state: string, action: any, fullState: any) => string;
        otherNameNoDefault: (state: any, action: any, fullState: any) => any;
        yetAnotherNameWithNullDefault: (state: string | null, action: any, fullState: any) => string | null;
    };
    selector: (state: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    selectors: {
        name: (state: any, props: any) => string;
        otherNameNoDefault: (state: any, props: any) => any;
        yetAnotherNameWithNullDefault: (state: any, props: any) => string | null;
    };
    values: {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
}`.trim(),
    )
})
