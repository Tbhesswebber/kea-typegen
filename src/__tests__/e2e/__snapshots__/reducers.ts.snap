// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers - as a function returning a object 1`] = `
"export interface logicType {
    actionCreators: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    actions: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    cache: Record<string, any>;
    path: [\\"logic\\"];
    pathString: \\"logic\\";
    props: Record<string, any>;
    reducer: (state: any, action: () => any, fullState: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    reducers: {
        name: (state: string, action: any, fullState: any) => string;
        otherNameNoDefault: (state: any, action: any, fullState: any) => any;
        yetAnotherNameWithNullDefault: (state: string | null, action: any, fullState: any) => string | null;
    };
    selector: (state: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    selectors: {
        name: (state: any, props: any) => string;
        otherNameNoDefault: (state: any, props: any) => any;
        yetAnotherNameWithNullDefault: (state: any, props: any) => string | null;
    };
    values: {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    _isKea: true;
}"
`;

exports[`reducers - as an object 1`] = `
"export interface logicType {
    actionCreators: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    actions: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    cache: Record<string, any>;
    path: [\\"logic\\"];
    pathString: \\"logic\\";
    props: Record<string, any>;
    reducer: (state: any, action: () => any, fullState: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    reducers: {
        name: (state: string, action: any, fullState: any) => string;
        otherNameNoDefault: (state: any, action: any, fullState: any) => any;
        yetAnotherNameWithNullDefault: (state: string | null, action: any, fullState: any) => string | null;
    };
    selector: (state: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    selectors: {
        name: (state: any, props: any) => string;
        otherNameNoDefault: (state: any, props: any) => any;
        yetAnotherNameWithNullDefault: (state: any, props: any) => string | null;
    };
    values: {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    _isKea: true;
}"
`;

exports[`reducers - with a function 1`] = `
"export interface logicType {
    actionCreators: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    actions: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    cache: Record<string, any>;
    path: [\\"logic\\"];
    pathString: \\"logic\\";
    props: Record<string, any>;
    reducer: (state: any, action: () => any, fullState: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    reducers: {
        name: (state: string, action: any, fullState: any) => string;
        otherNameNoDefault: (state: any, action: any, fullState: any) => any;
        yetAnotherNameWithNullDefault: (state: string | null, action: any, fullState: any) => string | null;
    };
    selector: (state: any) => {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    selectors: {
        name: (state: any, props: any) => string;
        otherNameNoDefault: (state: any, props: any) => any;
        yetAnotherNameWithNullDefault: (state: any, props: any) => string | null;
    };
    values: {
        name: string;
        otherNameNoDefault: any;
        yetAnotherNameWithNullDefault: string | null;
    };
    _isKea: true;
}"
`;

exports[`reducers - with bool default 1`] = `
"export interface logicType {
    actionCreators: {
        updateName: () => ({
            type: \\"update name (logic)\\";
            payload: {
                value: boolean;
            };
        });
    };
    actions: {
        updateName: () => ({
            type: \\"update name (logic)\\";
            payload: {
                value: boolean;
            };
        });
    };
    cache: Record<string, any>;
    path: [\\"logic\\"];
    pathString: \\"logic\\";
    props: Record<string, any>;
    reducer: (state: any, action: () => any, fullState: any) => {
        name: boolean;
    };
    reducers: {
        name: (state: boolean, action: any, fullState: any) => boolean;
    };
    selector: (state: any) => {
        name: boolean;
    };
    selectors: {
        name: (state: any, props: any) => boolean;
    };
    values: {
        name: boolean;
    };
    _isKea: true;
}"
`;
